src/passes/OptimizeInstructions.cpp

```diff
@@ -458,6 +458,90 @@ struct OptimizeInstructions
           return c;
         }
       }
+      {
+        Const* c;
+        Binary* bin;
+        Expression *x, *y;
+        BinaryOp op;
+
+        // fneg(x) * fneg(y)   ==>   x * y
+        // fneg(x) / fneg(y)   ==>   x / y
+        if (matches(
+              curr,
+              binary(&bin, Mul, unary(Neg, any(&x)), unary(Neg, any(&y)))) ||
+            matches(
+              curr,
+              binary(&bin, DivS, unary(Neg, any(&x)), unary(Neg, any(&y))))) {
+          bin->left = x;
+          bin->right = y;
+          return bin;
+        }
+        // fneg(x * C)   ==>   x * -C
+        // fneg(x / C)   ==>   x / -C
+        // fneg(C / x)   ==>   -C / x
+        //  if x != C and C != NaN
```

Comment:
why not allow a `NaN` here? I think it should work. If `C` is a NaN` then `x * C` is an arithmetic NaN with arbitrary sign, which the `fneg` flips, but that's ok.

Reply:
Right! Since I apply this only for `fastMath` we could skip this restriction

Reply:
To be clear, I think this is ok without fastMath.

Reply:
unfortunately it cause to fuzz issue: https://github.com/WebAssembly/binaryen/pull/3330#issuecomment-729954466

Reply:
(see what I said earlier for why I think it's ok without fastMath. it's nondeterministic, but it's ok - it's the same nondeterministic set of valid outputs, unless I'm missing something)

Reply:
Can you maybe post the fuzz testcase in nested mode (which is easier to read), and the output after optimization also in nested mode? (if you don't paste the output as well, then I need to either guess at what went wrong, or I need to build this branch and run it)

Reply:
I think it should be under `fastMath`. I compared Clang and GCC and GCC which usually more stricter don't apply similar simplification: https://godbolt.org/z/ExGq5W
while Clang do this.

Reply:
minimal original example:
```wat
(func $fn (result f32)
  (f32.neg
    (f32.mul
      (f32.const nan:0x7fff82)
      (f32.const 1.0)
    )
  )
)
```
optimized:
```wat
(func $fn (result f32)
  (f32.mul
    (f32.const nan:0x7fff82)
    (f32.const -1.0)
  )
)
```

Reply:
Here fiddle which demonstrate changed NaN sign before / after optimization:
https://webassembly.studio/?f=6hnx8occ068

Reply:
Thanks for the testcase. I think that is a valid optimization, but it's a limitation in the fuzzer that it can't allow it.

The set of valid outputs in both of those expressions is identical (the set of all arithmetic NaNs, of either sign). But we canonicalize NaNs in the interpreter to a positive NaN, so these end up different.

As a workaround I guess we can put such things behind `fastMath`, but ideally we'd find a way to fix the fuzzer. I don't have a good idea for how to do that atm. For now, how about putting a comment on such optimizations? like `This is valid even without `fastMath` but the fuzzer can be confused by it.`

Reply:
As I remember it's not valid without `fastMath`. For example spec test have test which disallow `x * -1` -> `-x` transformation due to `x * -1` will produce unsigned arithmetic NaN while `-x` will just change sign of NaN. It seems to me a very similar situation here

Reply:
`x * -1 -> -x` is not valid because the NaN propagation of `*` is different from the NaN propagation of `-`. Specifically, if there is a NaN input, then `*` can emit any arithmetic NaN, of any sign (a pretty big group of NaNs - anything where the payload starts with `1` basically; all other bits are free). But `-` must emit the exact same NaN just with sign flipped. So the spec is right to check that.

But the optimization here is valid, unless I'm missing something. If an input is a NaN, than `-(x*C)` can be any arithmetic NaN of any sign - because the inner multiply emits that, and flipping the sign on that set of NaNs leaves the set unchanged (`{ -x for x in S} === {x for x in S}` given that S is symmetric with regards to NaNs, if `x` is in `S` then so is `-x`). And on the other hand, `x*-C`'s valid outputs are that exact same set of all arithmetic NaNs (the sign of the constant doesn't matter; once `x` is a NaN, the output is that set of all arithmetic NaNs of both signs).

Reply:
Alright, I'll add TODO comment that mentions that at this stage we have to hide it under fastMath for pass fuzzer